CMSgov/onemac
# [DELETE] Oy2 7005 user lambda

Story: https://qmacbis.atlassian.net/browse/OY2-7005

Endpoint URL: https://d8a1yqw4i7q5r.cloudfront.net (Lambda needs to be tested in the console)

**Changes:**
Created a lambda to create/update User record or make a status change and send a confirmation email to their respective recipients

**Testing:**

Below are the request payloads for different use cases.

Testing tips:
Testing using the Lambda console:

- Login into the AWS Lambda console
- Select the lambda: oy2-7005-user-lambda-putUser
- Testing using Postman:
- If you choose to use postman, then remove the "body:{" wrapper from the payloads listed below as Postman would automatically add that for you.
- You may want to open the DynamoDb console in a browser to watch the data changes and cleanup as needed, 
- Cloudwatch console to watch for validation, and success and logs
- Use different test email ids for different roles and keep the inboxes open to check for the email confirmations

General Info:

The property isPutUser is used to set the intent of this lambda call. Two possible intents are;
  1. isPutUser must be set to true, to Insert a new user record or edit most of the properties of an existing user such as first or last name or to change the status
  2. isPutUser must be set to false or not included as a property at all for status change mode. This helps just to change the status of a user.
  3. To over-ride the existing system admin email, a substitute email id needs to be set in the AWS Lambda console > configuration > environment variables > systemAdminEmailId. Otherwise, system admin emails would go to Sabrina’s inbox

Sample putUser payload:
```
{
	"userEmail": “cmsapprover@somedomain.com”,
	"doneBy": “systemadmin@somedomain.com",
	"attributes": [
		{
			"status": "pending" 
			// “stateCode”: “MD”.  // need to include stateCode property if the type is a state user or state admin
			}
	],
	"isPutUser": true,
	"firstName": “John”, 
	"lastName": "Doe",
	"type": "cmsapprover" 
}
```

Sample status change payload:

```
{
	"userEmail": "stateadmin@somedomain.com",
	"doneBy": "cmsapprover@somedomain.com",
	"attributes": [
		{
			“stateCode: “MD”,  // required for state user and state admin
			"status": "pending"
		}
	],
	“type”:  “stateadmin”
}
```

**Test Cases:**

**1)Common validations:**

1.1. userEmail is required and must be a valid email id
1.2. doneBy is required and must be a valid email id
1.3, attributes is required
1.4  attributes status can only be one of 'pending', 'denied', 'revoked', 'active'
1.5  if user type is state user or state admin then the attributes must contain stateCode and it must a valid state code

**2) Put User:**

2.1.  With isPutUser set to true;
	i) Should see a validation error if the type is missing or not in (’stateuser', 'stateadmin', 'cmsapprover’)
	ii) Should see a validation error if firstName or last name is missing when the user record does not previously exist (insert mode). They are optional in case of updating an existing user.
2.2.  Should see a validation error; If attributes objects status is any thing except ‘pending’

**3. Common validations for both PutUser and statusChange:**

    3.1 doneBy user type must be one level higher in hierarchy. Example: Only a state admin can change the status of a state user
    3.2 Disallowed status changes: 
          - can't go from denied to revoked or revoked to denied
          - active cannot go to denied, only pending can go to denied?
          - pending cannot go to revoked, revoked is only from active (or granted)
          - any status cannot go to pending

**4 Email:**

    4.1 Confirmation email needs to be sent to;

            1.  state user status change confirmation email to the respective state admins
            2.  state admin status change confirmation email to the cmsapprover
            3.  cmsapprover status change confirmation email to the system admins
            4.  If the system admin email provisioned as env variable, then that should over ride all the system admin email ids retrieved from the database.

--- Happy Testing :) ---
